### PR TITLE
User mentions

### PR DESCRIPTION
**Pull request by Stvad**
LMK if there is a better way to deal with ck-editor reset/specificity stuff!

image
image

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203709701440261) by [Unito](https://www.unito.io)

Remaining TODOs:
- [x] Test email notifications (still figuring out how to do that locally)
- [ ] Various comments in the code that I have questions about

Follow-up:
- [ ] User can receive 2 notifications (when they are tagged on a comment in the post they are subscribed to)
- [ ] Right now the limit on number of mentions is not exposed to the user, so tagging with silently fail

### CODE REVIEW

File: packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx
Comment (Raemon) [inline packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx:31]: This should use `userGetDisplayName`, from users/helpers.tsx. (This is pretty non-obvious, at some point we should figure out a good eslint setup that warns when you use certain fields that are actually a bit of a footgun)

Comment (Stvad) [reply packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx:31]: Will update! curious what is the footgun around that field? is it missing sometimes/something else?

File: packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx
Comment (Stvad) [inline packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx:29]: note to self: the `ck-override` class is duplicated

File: packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx
Comment (Stvad) [reply packages/lesswrong/components/search/UsersSearchAutocompleteHit.tsx:31]: ok, looking at it seems alignment forum has a different notation, which probably implies that I should retrieve `fullName` from search above as-well

**Comment by Stvad** [comment]
Starting to push backend changes - lmk if you'd prefer me to split it out into a different PR!

**Comment by Raemon** [comment]
I think it's best to continue with this PR, rather than making another one. (I'm a bit on the fence. It might be nice for smaller-review-size to make a PR that you merge _into_ this PR and let this PR become the feature-branch. But merging this without the rest of the job done doesn't seem worth it)

**Comment by jpaddison3** [comment]
Sorry about this, this was an automation

image

In general though, could you make this a draft PR?

File: packages/lesswrong/lib/collections/comments/collection.ts
Comment (Raemon) [inline packages/lesswrong/lib/collections/comments/collection.ts:100]: I think having mentions in comments and posts covers the use cases that matter most. I think we want them in tag _discussions_, but not tags themselves, and I think tag-discussions just use comments?

(We may not have pingbacks enabled in comments, in which case I definitely agree we'd want to turn that on and/or figure out what to do instead)

Comment (Stvad) [reply packages/lesswrong/lib/collections/comments/collection.ts:100]: this line enables the pingbacks in the comments. and indeed the tag-discussions are just comments - I've tested doing mentions on those.

File: packages/lesswrong/lib/notificationTypes.tsx
Comment (b0b3rt) [inline packages/lesswrong/lib/notificationTypes.tsx:83]: 🎉

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (b0b3rt) [inline packages/lesswrong/server/editor/mentions-notify.ts:14]: nit: prefer `??` when not explicitly relying on `||` to coalesce on `0`/`''`/other non-null/undefined false-y values

File: packages/lesswrong/server/notificationTypesServer.tsx
Comment (b0b3rt) [inline packages/lesswrong/server/notificationTypesServer.tsx:606]: The cast seems wrong, I think it should be on `documentType`?  `documentId` is the actual `_id` of the document being referenced by the notification

File: packages/lesswrong/server/notificationTypesServer.tsx
Comment (b0b3rt) [inline packages/lesswrong/server/notificationTypesServer.tsx:613]: Same as https://github.com/ForumMagnum/ForumMagnum/pull/6458/files#r1080768601

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:14]: 👍 should we enable https://typescript-eslint.io/rules/prefer-nullish-coalescing/ ?

File: packages/lesswrong/server/notificationTypesServer.tsx
Comment (Stvad) [reply packages/lesswrong/server/notificationTypesServer.tsx:606]: whelp 🙈. nice catch!

Comment (Stvad) [reply packages/lesswrong/server/notificationTypesServer.tsx:606]: I suppose I'd ideally propagate the type to the DB (update allowed values, etc) but when I looked at it - it felt like that refactoring was expanding too much in scope 🤔

File: packages/lesswrong/lib/collections/comments/collection.ts
Comment (Stvad) [reply packages/lesswrong/lib/collections/comments/collection.ts:100]: To confirm and resolve this one: for now we are ok with mentions working only on posts and comments?

Comment (Raemon) [reply packages/lesswrong/lib/collections/comments/collection.ts:100]: Yes.

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (Stvad) [inline packages/lesswrong/server/editor/mentions-notify.ts:40]: special limit for admins: 50

**Comment by jpaddison3** [comment]
I'm still going to wait on @Raemon's review before assigning @s-cheng.

File: packages/lesswrong/lib/notificationTypes.tsx
Comment (Stvad) [inline packages/lesswrong/lib/notificationTypes.tsx:90]: todo: add more on why

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (Raemon) [inline packages/lesswrong/server/editor/mentions-notify.ts:12]: Give this a name that more clearly says what it's for (i.e. notifyUsersAboutMentions)

File: packages/lesswrong/lib/notificationTypes.tsx
Comment (Raemon) [reply packages/lesswrong/lib/notificationTypes.tsx:90]: @jimrandomh seemed probably good for you to take a look at this refactor and see if it made sense to you.

File: packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts
Comment (Raemon) [inline packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts:19]: @jpaddison3 or @oetherington just doublechecking this is done right.

Comment (jpaddison3) [reply packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts:19]: * These are migrations to be run on a postgres database, and will indeed not be run on any non-postgres database.
* They need to update the postgres schema, which this doesn't do. See, for example: packages/lesswrong/server/migrations/20230118T104229.add_subforumIntroPostId.ts
* You'll want to run a "manualMigrations" for a mongodb migration. That one can run the fillDefaultValues code.
* The fillDefaultValues code is still valuable here though, after the addField code.

@oetherington can correct me if I got any of that wrong.

**Comment by Raemon** [comment]
I've now done a first pass review, talking it over with Vlad and testing out locally. I left a couple comments for @jimrandomh and EA Forum people, re: some domains I wasn't that confident in. (namely: "how exactly do migrations work? Did Vlad handle it right?" and "was vlad's refactoring/restructuring of some code in the notifications system a net-benefit)

I left a couple other individual comments. I think it's possible @oetherington should take a look at the migrations stuff.

File: packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts
Comment (oetherington) [reply packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts:19]: Yep, as JP said, you'll need to `await addField(db, Users, "notificationNewMention")`, and also `await addField(db, Comments, "pingbacks")`. Ideally, you should also put calls to `dropField` in the down migration too (the file JP mentioned has exactly the same pattern here too).

The `fillDefaultValues` is redundant here and can be safely removed. The default value is built into the column type in postgres, so it will be automatically filled when the column is created.

Comment (Stvad) [reply packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts:19]: I see. So as the things stand currently - I need to create both manual migrations (to update mongo) and Postgres migration in a pattern you describe, right?

And looking at the recent Slack discussion - Postgres migration would happen automatically & manualMigration would need to be run manually on prod before pushing this PR

File: packages/lesswrong/lib/notificationTypes.tsx
Comment (Stvad) [reply packages/lesswrong/lib/notificationTypes.tsx:90]: Motivation for this was that:
- I needed to handle the case where I have different document types for single notification type, so I need a function like this (though not necessarily covering all the cases)
- Most other notification types follow the pattern of get the title of the thing and the author of the thing and can be rewritten using this function. 
  - Which abstracts away the details of the correct way of getting title/author (rn each notification has to know about and use the proper utility functions for that e.g. using `userGetDisplayName` vs using `.displayName` as a few of the types do rn)
  - Actually probably most of "getMessage" functions can be auto-generated from a template string (with author and title being template parameters)

Plausibly getLink stuff should also be incorporated here to minimize number of times we do switch over this set of values, but that was a different code flow that I preferred not to touch for now.

File: packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts
Comment (oetherington) [reply packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts:19]: Yeah - unfortunately that's the case at the moment while we're switching over. It should be simpler once the process is finished.

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:40]: actually, I think I should probably just allow the mentions for admins vs having a specific arbitrary threshold

File: packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts
Comment (Stvad) [reply packages/lesswrong/server/migrations/20230119T202810.user_mentions_newMentionNotifictions.ts]: thanks for explanations! updated the migration code!

**Comment by Xodarap** [comment]
Minor: you can mention yourself. Probably fine?

**Comment by Xodarap** [comment]
When you mention someone in a draft they get notified

**Comment by Stvad** [comment]
> Minor: you can mention yourself. Probably fine?

you can @tag yourself in UI, but won't receive notification, which is based on my intuition of this should work..

**Comment by Stvad** [comment]
> When you mention someone in a draft they get notified

Ah, let me see if I understand drafts! 
- only DbPost can be a draft
- Draft posts are only viewable by the author and users they've shared the draft with.

Variations on expected behavior:
1. Notifications are only delivered when the draft is published.
2. Notifications are immediately sent to users who have access to the document at the draft stage. 
  When draft is published notifications are delivered to other mentioned users

Does that make sense?

**Review by Xodarap — APPROVED** [review]
I think you should fix the draft thing and the migration thing. Otherwise I left some small comments which are optional to fix. Thanks for doing this!

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:19]: Yeah, that's definitely not an ideal UX! I noted that down in the PR description as a follow-up work

**Comment by Raemon** [comment]
Ah, good catch with the drafts-issues. I agree with the summary of "notify users who are shared on it immediately, notify everyone else when the post was published."

File: packages/lesswrong/server/editor/mentions-notify.ts
Comment (Raemon) [reply packages/lesswrong/server/editor/mentions-notify.ts:19]: It occurs to me, maybe it's just better to leave out this restriction until we've actually built good UI to warn about it. (I don't expect it to get abused immediately, and we can avoid publicizing the feature until the restrictions are built)

(alternately, if there's anything we can do to throw an error here without too much additional dev work, that'd probably be best for now)

Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:19]: Plausibly literally throwing exception should be ok. (from what I saw elsewhere that is handled relatively gracefully)

Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:19]: image
yeah, seems ok

Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:20]: Though actually to make that style of error work properly one would need to move the callbacks into the "before" lifecycle instead of after - otherwise weird things happen. 
it seems fine to move it to before though

Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:20]: Hrm actually moving to "before" has issues 🤔

Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:20]: So the behavior if I were to keep the callback in "after" & throw exception is something like:
- operation gets performed (comment created, post created or updated)
- User will see the exception about pingbacks and not get redirected (so the perception is that operation didn't work) 
- presumably none of the further "after" callbacks got run. 

On balance, even though it's giving user clearer feedback, it adds it's own confusion and bug potential and I wouldn't go there

Comment (Stvad) [reply packages/lesswrong/server/editor/mentions-notify.ts:20]: as an 80/20 thing set the limit to be 25 to start with - and we can lower it when better ui support is available

**Comment by Stvad** [comment]
Can one leave comments on a draft? 

On testing - comment field doesn't show up, though I thought I remember seeing on a prev draft, though might be imagining it.

Also it seems side-comments have a different editor and so don't support the mentions atm @Raemon

**Comment by Raemon** [comment]
Will look over tomorrow. @Xodarap I think in Slack Vlad had asked if you could take a look at how it handles drafts. I'll also look at it but seemed like you might want to give it another look.

**Comment by Raemon** [comment]
> Also it seems side-comments have a different editor and so don't support the mentions atm

By side comments do you mean the inline-comments in the editor (yes I agree these don't support mentions without a lot more work), or the newer side comments in the main post page (you can highlight text to make a comment which is retroactively treated as a sort of side-comment)? (the latter I expect should work fine)

**Comment by Stvad** [comment]
> > Also it seems side-comments have a different editor and so don't support the mentions atm
> 
> By side comments do you mean the inline-comments in the editor (yes I agree these don't support mentions without a lot more work), or the newer side comments in the main post page (you can highlight text to make a comment which is retroactively treated as a sort of side-comment)? (the latter I expect should work fine)

Oh, yes, you're right I was confused about that distinction!

File: packages/lesswrong/server/editor/make_editable_callbacks.ts
Comment (Raemon) [inline packages/lesswrong/server/editor/make_editable_callbacks.ts:7]: Note: I think @Discordius (Oliver Habryka) actually prefers the one-line rather than 5-line version of this. (Oli's less involved with the codebase now, and I'm not sure what other people think. I used to go around changing the codebase in the way you're doing here and Oli trained it out of me when he was the primary decisionmaker). 

His reasoning was something like "this line isn't actually that important, and having it take up more vertical space assigns the line disproportionate importance, and meanwhile parsing out the different variables isn't useful enough to counteract that".

There's a few other places you do this, I'm not personally gonna ask you to change them, but noting it's at least not considered by everyone to be a strict improvement.

Comment (Stvad) [reply packages/lesswrong/server/editor/make_editable_callbacks.ts:7]: This is not exactly an intentional change (though it matches my preferences) - my editor does the  automatically when it auto-imports things for you/etc.  I mostly tried to avoid reformatting code I didn't touch,  but didn't make effort to undo things like this

**Comment by Raemon** [comment]
I tested that the undrafting process worked. Seemed to work fine. Approved.

**Review by Raemon — APPROVED** [review]

**Pull request merged**
